rust/noosphere-core/src/bench.rs

```diff
@@ -0,0 +1,182 @@
+        ms_len = std::cmp::max(ms_len, digit_len(stat.median_duration_ms));
+        bytes_len = std::cmp::max(bytes_len, digit_len(stat.median_bytes));
+    }
+
+    println!("");
+    println!(
+        "{} {} {} {}",
+        pad_string("name", name_len, false),
+        pad_string("iters", iters_len, false),
+        pad_string("ms", ms_len, false),
+        pad_string("bytes", bytes_len, false)
+    );
+    for stat in stats {
+        println!(
+            "{} {} {} {}",
+            pad_string(&stat.name, name_len, false),
+            pad_string(&stat.iterations.to_string(), iters_len, true),
+            pad_string(&stat.median_duration_ms.to_string(), ms_len, true),
+            pad_string(&stat.median_bytes.to_string(), bytes_len, true)
+        );
+    }
+}
+
+async fn run_benchmark(bench_desc: BenchDesc) -> Result<BenchStats> {
+    let mut benchers = vec![];
+    for _ in 0..bench_desc.iterations {
+        let mut bs = Bencher::default();
+        let _ = (bench_desc.func)(&mut bs).await;
+        benchers.push(bs);
+    }
+
+    Ok(BenchStats::from((bench_desc, benchers)))
+}
+
+pub async fn run_benches(benches: Vec<BenchDesc>) -> Result<()> {
+    let mut benchmark_results = vec![];
+    for bench in benches {
+        let benchers = run_benchmark(bench).await?;
+        benchmark_results.push(benchers);
+    }
+    print_bench_stats(benchmark_results);
+    Ok(())
+}
+
+#[macro_export]
+macro_rules! noosphere_benchmark {
+    ($(($function:ident, $iterations:expr)),+) => {
+        fn noosphere_benchmark_collect_benches() -> ::std::vec::Vec<$crate::bench::BenchDesc> {
+            use $crate::bench::BenchDesc;
+            use ::std::borrow::Cow;
+            use ::std::boxed::Box;
+            let mut benches = ::std::vec::Vec::new();
+            $(
+                benches.push(BenchDesc {
+                    name: Cow::from(stringify!($function)),
+                    ignore: false,
+                    func: Box::new(move |mut bencher: &mut Bencher| {
+                        Box::pin($function(bencher))
+                    }),
+                    iterations: $iterations,
+                });
+            )+
+            benches
+        }
+
+        #[cfg(target_arch = "wasm32")]
+        use wasm_bindgen_test::wasm_bindgen_test;
+        #[cfg(target_arch = "wasm32")]
+        wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
+
+        #[cfg(target_arch = "wasm32")]
+        pub fn main() {}
+        #[cfg_attr(target_arch = "wasm32", wasm_bindgen_test)]
+        async fn noosphere_benchmark_wasm32_main() {
+            $crate::tracing::initialize_tracing(None);
+            let benches = noosphere_benchmark_collect_benches();
+            $crate::bench::run_benches(benches).await.unwrap();
+        }
+
+        #[cfg(not(target_arch = "wasm32"))]
+        #[tokio::main(flavor = "multi_thread")]
+        pub async fn main() {
```

Comment:
Why is it that non-Wasm is a `main` entrypoint while Wasm is a test?

Reply:
wasm_bindgen_test is the "host" executable; can run in browsers, though limited on headless due to timeouts

Reply:
I guess what I'm reaching for is: do we need asymmetry here, or can we structure the "native" branch as a test also (thereby using a combined entrypoint for both cases)?